app/breadcrumb.rb

```diff
@@ -4,145 +4,63 @@ class BreadcrumbGenerator
     @type = type
+    @unitID = (type == "unit") ? thisPageName : getItemsUnit(thisPageName) 
   end
 
-  # 1) Generate an array of nodes containing name/url pairs
+  # Returns an array containing topmost item under root ID and name, or empty array if root
+  def getCampusInfo
+    unitID = @unitID
+    if unitID == 'root' 
+      return [] 
+    else
+      until isCampus?(unitID) 
+        unitID = UnitHier.filter(:unit_id => unitID, :is_direct => true).order(:ordering).map(:ancestor_unit)[0]
+      end        
+    end
+    unit = Unit[unitID]
+    return [unitID, unit.name]
+  end
+
+  # Generate an array of hashes containing name/url key-values 
   # If it's unit level, find parent using unit_hier table, otherwise,
   # for item level page look for the parent unit using the unit_item table
-  # 2) Feed nodes into breadcrumb class
-  def generate
+  def generateCrumb
     nodes = []
 
-    if @type == "unit"
-      unitID = @thisPageName
-    else # type = item 
+    if @type == "item"
       # ToDo: Journal/Non-journal displays differently
-      nodes << ["Volume/Issue placeholder"]
-      itemID = 'qt' + @thisPageName
-      unitID = UnitItem.filter(:item_id => itemID, :is_direct => true).map(:unit_id)[0]
+      nodes.unshift({"name" => "Volume/Issue placeholder", "url" => "/item/#{@thisPageName}"})
     end
 
+    unitID = @unitID
     until unitID == "root"
       unit = Unit[unitID]
-      nodes << [unit.name, unitID != @thisPageName? "/unit/#{unitID}" : nil]
+      nodes.unshift({"name" => unit.name, "url" => "/unit/#{unitID}"})
       # ToDo: Handle mutiple campuses
       unitID = UnitHier.filter(:unit_id => unitID, :is_direct => true).order(:ordering).map(:ancestor_unit)[0]
     end
-    nodes << ["eScholarship"]
-
-    breadcrumbs = Breadcrumbs.new
-    nodes.reverse_each {|n| breadcrumbs.add n[0], n[1]}
-    breadcrumbs.render(format: "Inline")
-  end
-
-end
-
-###################################################################################################
-# Breadcrumbs functionality courtesy https://github.com/fnando/breadcrumbs
-#
-# Implement the Render:Base class differently to generate a different navigation style. Currently 'Inline'
-
-class Breadcrumbs
-  attr_accessor :items
-
-  def initialize
-    self.items = []
+    nodes.unshift({"name" => "eScholarship", "url" => "/"})
+    return nodes
   end
 
-  # Add a new breadcrumbs.
-  #
-  #   breadcrumbs.add 'Home'
-  #   breadcrumbs.add 'Home', '/'
-  #   breadcrumbs.add 'Home', '/', class: 'home'
-  #
-  def add(text, url = nil, options = {})
-    # options = {i18n: true}.merge(options)
-    # text = translate(text) if options.delete(:i18n)
-    items << [text.to_s, url, options]
+  def getItemsUnit(pageName)
+    itemID = 'qt' + pageName
+    unitID = UnitItem.filter(:item_id => itemID, :is_direct => true).order(:ordering_of_units).map(:unit_id)[0]
+    if !unitID
+      pp("No parent Unit for this item ")
+      raise "No Unit for this item " + itemID
```

Comment:
My exceptions are not being displayed in the final 404 displayed in the UI, so I need to look further into why this is broken.